diff -NBaur emulationstation-de-stable-3.0-AUR/CMake/Packages/FindPoppler.cmake emulationstation-de-stable-3.0-AUR-retroroot/CMake/Packages/FindPoppler.cmake
--- emulationstation-de-stable-3.0-AUR/CMake/Packages/FindPoppler.cmake	2024-03-16 13:32:03.000000000 +0100
+++ emulationstation-de-stable-3.0-AUR-retroroot/CMake/Packages/FindPoppler.cmake	2024-04-23 13:16:41.509235307 +0200
@@ -87,9 +87,13 @@
   # Scan poppler libraries for dependencies on Fontconfig
   include(GetPrerequisites)
   mark_as_advanced(gp_cmd)
-  GET_PREREQUISITES("${POPPLER_LIBRARY}" POPPLER_PREREQS 1 0 "" "")
-  if("${POPPLER_PREREQS}" MATCHES "fontconfig")
-    set(POPPLER_NEEDS_FONTCONFIG TRUE)
+  if( NOT RETROROOT )
+    GET_PREREQUISITES("${POPPLER_LIBRARY}" POPPLER_PREREQS 1 0 "" "")
+    if("${POPPLER_PREREQS}" MATCHES "fontconfig")
+      set(POPPLER_NEEDS_FONTCONFIG TRUE)
+    else()
+      set(POPPLER_NEEDS_FONTCONFIG FALSE)
+    endif()
   else()
     set(POPPLER_NEEDS_FONTCONFIG FALSE)
   endif()
