diff -urN SDL2-2.30.11-ori/src/video/kmsdrm/SDL_kmsdrmvideo.c SDL2-2.30.11/src/video/kmsdrm/SDL_kmsdrmvideo.c
--- SDL2-2.30.11-ori/src/video/kmsdrm/SDL_kmsdrmvideo.c	2025-01-01 16:48:45.000000000 +0100
+++ SDL2-2.30.11/src/video/kmsdrm/SDL_kmsdrmvideo.c	2025-01-14 15:49:39.570349189 +0100
@@ -363,6 +363,7 @@
        Use of the modifiers is necessary on some platforms. */
     w = KMSDRM_gbm_bo_get_width(bo);
     h = KMSDRM_gbm_bo_get_height(bo);
+#if 0
     format = KMSDRM_gbm_bo_get_format(bo);
 
     if (KMSDRM_drmModeAddFB2WithModifiers &&
@@ -387,11 +388,12 @@
 
         rc = KMSDRM_drmModeAddFB2WithModifiers(viddata->drm_fd, w, h, format, handles, strides, offsets, modifiers, &fb_info->fb_id, flags);
     }
+#endif
 
     if (rc < 0) {
         strides[0] = KMSDRM_gbm_bo_get_stride(bo);
         handles[0] = KMSDRM_gbm_bo_get_handle(bo).u32;
-        rc = KMSDRM_drmModeAddFB(viddata->drm_fd, w, h, 24, 32, strides[0], handles[0], &fb_info->fb_id);
+        rc = KMSDRM_drmModeAddFB(viddata->drm_fd, w, h, 16, 16, strides[0], handles[0], &fb_info->fb_id);
     }
 
     if (rc < 0) {
@@ -512,7 +514,7 @@
 
     target.w = width;
     target.h = height;
-    target.format = 0; /* Will use the default mode format. */
+    target.format = SDL_PIXELFORMAT_RGB565; /* Will use the default mode format. */
     target.refresh_rate = refresh_rate;
     target.driverdata = 0; /* Initialize to 0 */
 
@@ -901,7 +903,7 @@
     display.desktop_mode.w = dispdata->mode.hdisplay;
     display.desktop_mode.h = dispdata->mode.vdisplay;
     display.desktop_mode.refresh_rate = dispdata->mode.vrefresh;
-    display.desktop_mode.format = SDL_PIXELFORMAT_ARGB8888;
+    display.desktop_mode.format = SDL_PIXELFORMAT_RGB565;
     display.desktop_mode.driverdata = modedata;
     display.current_mode = display.desktop_mode;
 
@@ -1193,7 +1195,7 @@
     SDL_VideoDisplay *display = SDL_GetDisplayForWindow(window);
     SDL_DisplayData *dispdata = (SDL_DisplayData *)display->driverdata;
 
-    uint32_t surface_fmt = GBM_FORMAT_ARGB8888;
+    uint32_t surface_fmt = GBM_FORMAT_RGB565;
     uint32_t surface_flags = GBM_BO_USE_SCANOUT | GBM_BO_USE_RENDERING;
 
     EGLContext egl_context;
@@ -1219,7 +1221,7 @@
     display->current_mode.w = dispdata->mode.hdisplay;
     display->current_mode.h = dispdata->mode.vdisplay;
     display->current_mode.refresh_rate = dispdata->mode.vrefresh;
-    display->current_mode.format = SDL_PIXELFORMAT_ARGB8888;
+    display->current_mode.format = SDL_PIXELFORMAT_RGB565;
 
     windata->gs = KMSDRM_gbm_surface_create(viddata->gbm_dev,
                                             dispdata->mode.hdisplay, dispdata->mode.vdisplay,
@@ -1366,7 +1368,7 @@
         mode.w = conn->modes[i].hdisplay;
         mode.h = conn->modes[i].vdisplay;
         mode.refresh_rate = conn->modes[i].vrefresh;
-        mode.format = SDL_PIXELFORMAT_ARGB8888;
+        mode.format = SDL_PIXELFORMAT_RGB565;
         mode.driverdata = modedata;
 
         if (!SDL_AddDisplayMode(display, &mode)) {
